### PR TITLE
[runtime] Throw initialization EE exceptions, don't assert

### PR DESCRIPTION
**Pull request by alexanderkyte**

### CODE REVIEW

**Comment by alexanderkyte** [comment]
@lambdageek does this look like correct API usage to you? It works experimentally.

**Comment by lambdageek** [comment]
Yea that looks ok, although I'm not familiar with mini_llvmonly_initial_imt_tramp so can't vouch that it makes sense for it to throw.

File: mono/mini/jit-icalls.c [hunk 1]
```diff
@@ -1635,7 +1635,9 @@ resolve_vcall (MonoVTable *vt, int slot, MonoMethod *imt_method, gpointer *out_a
 
 	// FIXME: This can throw exceptions
 	addr = compiled_method = mono_compile_method_checked (m, error);
-	mono_error_assert_ok (error);
+	if (!is_ok (error))
```
Comment (vargaz) [inline mono/mini/jit-icalls.c]: This can be return_val_if_nok (NULL, error);

**Comment by vargaz** [comment]
Looks ok otherwise.

**Comment by alexanderkyte** [comment]
What this basically means is that if we run into an ExecutionEngine exception when initializing the vtable slots, we will throw the ExecutionEngine exception. This is important because it allows us to catch failures during CI runs to provide useful reporting. I'm not sure if there's a more appropriate place, but it's important to make it a prompt exception. 

If we have a pending exception and try to call the constructor for an exception, the cooperative check gets confused for a failure in the actual runtime invoke. Delayed ExecutionEngine exceptions were thus getting asserted on before even calling the constructor.

**Comment by alexanderkyte** [comment]
I've had to make other changes to ensure that early failures could be remembered. We keep the imt entry set to the default that results in another runtime invoke and a well-behaved exception.

**Comment by alexanderkyte** [comment]
Don't merge yet, I found another way for a failure here to assert. Need to add to the PR.

**Pull request closed without merging**